Add Message#getStartedThread and ThreadChannel#retrieveParentMessage

Almighty-Satan opened the pull request:
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

<!--
  There are several guidelines you should follow in order for your
  Pull Request to be merged.
-->

- [X] I have checked the PRs for upcoming features/bug fixes.
- [X] I have read the [contributing guidelines][contributing].

<!--
  It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.
-->

### Changes

- [x] Internal code
- [X] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ <!-- Insert other type here -->

<!-- Replace "NaN" with an issue number if this is a response to an issue -->

Closes Issue: #2092

## Description

Adds a `getStartedThread()` method to the Message object to provide a method that returns the Thread that was started from the message (if one exists).

Example:
```Java
Message message = //retrieve the message somehow
ThreadChannel thread = message.getStartedThread();
if (thread != null) {
    System.out.println(thread.getName());
    thread.sendMessage("Hello thread people!").queue();
}
```

Tais993 commented on src/main/java/net/dv8tion/jda/api/entities/Message.java:
```diff
@@ -2524,6 +2524,14 @@ default MessageReaction getReactionById(@Nonnull String id)
     boolean isEphemeral();
 
+    /**
+     * Returns a possibly null {@link net.dv8tion.jda.api.entities.ThreadChannel} that this message started
+     * 
+     * @return The {@link net.dv8tion.jda.api.entities.ThreadChannel} that this message started
```
```suggestion
     * Returns a possibly null {@link ThreadChannel} that this message started
     * 
     * @return The {@link ThreadChannel} that this message started
```

DV8FromTheWorld reviewed (commented):
Works for me. However, some internal conversation with @MinnDevelopment led us to think another name might be preferable.

Some ideas that come to mind (I am endorsing none of them yet 😋):
 - `getStartedThread`
 - `getSpawnedThread`
 - `getChildThread` (Might match a `retrieveParentMessage` in `ThreadChannel`?)

Almighty-Satan commented:
@DV8FromTheWorld I've added `ThreadChannel#retrieveParentMessage` and renamed `Message#getReferencedThreadChannel` to `getChildThread`. Although I personally think `getSpawnedThread` (or `getStartedThread`) would be a little less confusing.

DV8FromTheWorld commented:
@MinnDevelopment opinions on naming?
@Almighty-Satan they were ideas, not requirements. Happy to talk it out

MinnDevelopment commented:
I'd go with `getStartedThread`, it would match the endpoint name too: [Start Thread from Message](https://discord.com/developers/docs/resources/channel#start-thread-from-message)

DV8FromTheWorld commented:
Works for me.

DV8FromTheWorld commented on src/main/java/net/dv8tion/jda/api/entities/ThreadChannel.java:
```diff
@@ -140,6 +140,47 @@ default GuildMessageChannel getParentMessageChannel()
+    /**
+     * Attempts to get the {@link net.dv8tion.jda.api.entities.Message Message} from Discord's servers that started this thread.
+     *
+     * <p>The {@link Message#getMember() Message.getMember()} method will always return null for the resulting message.
+     * To retrieve the member you can use {@code getGuild().retrieveMember(message.getAuthor())}.
```
I know this comes from our other docs related to retrieveMessage, but I question it.
@MinnDevelopment is this actually valid anymore? Stepping through the logic 
 - [MessageChannel#retrieveMessageById uses EntityBuilder#createMessageWithChannel](https://github.com/DV8FromTheWorld/JDA/blob/b46ea213f3f6ac6daaf764b045a6aacd6e9865d1/src/main/java/net/dv8tion/jda/api/entities/MessageChannel.java#L839)
  - [EntityBuilder#createMessageWithChannel uses #createMessage0](https://github.com/DV8FromTheWorld/JDA/blob/d7bdd6c5f6ddddd5fa53a5853fde17f13bd3f163/src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java#L1331)
  - [EntityBuilder#createMessage0 doesn't build the member because the json doesn't container 'member'](https://github.com/DV8FromTheWorld/JDA/blob/d7bdd6c5f6ddddd5fa53a5853fde17f13bd3f163/src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java#L1401)
  - [EntityBuilder#createMessage0 attaches member information if it didn't have it but it exists in the cache](https://github.com/DV8FromTheWorld/JDA/blob/d7bdd6c5f6ddddd5fa53a5853fde17f13bd3f163/src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java#L1438-L1439)
  
  It seems like if Member is cached then a Rest retrieved message could have a member attached to it.

MinnDevelopment replied:
It is unreliable, so i just documented as null. Documenting all cache edge cases seems overkill.

DV8FromTheWorld commented on src/main/java/net/dv8tion/jda/api/entities/ThreadChannel.java:
```diff
@@ -140,6 +140,47 @@ default GuildMessageChannel getParentMessageChannel()
+     * @throws net.dv8tion.jda.api.exceptions.AccountTypeException
+     *         If the currently logged in account is not from {@link net.dv8tion.jda.api.AccountType#BOT AccountType.BOT}
+     * @throws net.dv8tion.jda.api.exceptions.InsufficientPermissionException
+     *         If this is a {@link net.dv8tion.jda.api.entities.TextChannel TextChannel} and the logged in account does not have
```
Doesn't need to just be a TextChannel. Any GuildMessageChannel that supports permissions has the opportunity to throw this

DV8FromTheWorld replied:
yes

Almighty-Satan replied:
Should I just do it in this PR or create a new one for that?

Almighty-Satan commented on src/main/java/net/dv8tion/jda/api/entities/Message.java:
```diff
@@ -2524,6 +2524,14 @@ default MessageReaction getReactionById(@Nonnull String id)
     boolean isEphemeral();
 
+    /**
+     * Returns a possibly null {@link ThreadChannel} that this message started
```
Is there a specific reason you put an `@code` tag around the second `null` but not the first one?

DV8FromTheWorld commented on src/main/java/net/dv8tion/jda/api/entities/Message.java:
```diff
@@ -2333,6 +2333,15 @@ default MessageReaction getReactionById(@Nonnull String id)
+     * Returns a possibly {@code null} {@link net.dv8tion.jda.api.entities.ThreadChannel ThreadChannel} that this message started.
+     * This can be {@code null} due to no ThreadChannel being started from it or the ThreadChannel later being deleted.
+     * 
+     * @return The {@link net.dv8tion.jda.api.entities.ThreadChannel ThreadChannel} that this message started
```
```suggestion
     * @return The {@link net.dv8tion.jda.api.entities.ThreadChannel ThreadChannel} that was started from this message.
```
